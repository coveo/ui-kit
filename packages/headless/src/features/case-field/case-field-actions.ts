import {createAction, createAsyncThunk} from '@reduxjs/toolkit';
import {getVisitorID} from '../../api/analytics/coveo-analytics-utils';
import {isErrorResponse} from '../../api/search/search-api-client';
import {AsyncThunkCaseAssistOptions} from '../../api/service/case-assist/case-assist-api-client';
import {prepareContextFromFields} from '../../api/service/case-assist/case-assist-params';
import {GetCaseClassificationsRequest} from '../../api/service/case-assist/get-case-classifications/get-case-classifications-request';
import {GetCaseClassificationsResponse} from '../../api/service/case-assist/get-case-classifications/get-case-classifications-response';
import {
  CaseAssistConfigurationSection,
  CaseFieldSection,
  CaseInputSection,
  ConfigurationSection,
  DebugSection,
} from '../../state/state-sections';
import {
  validatePayload,
  requiredNonEmptyString,
  requiredEmptyAllowedString,
} from '../../utils/validate-payload';

export interface SetCaseFieldActionCreatorPayload {
  /**
   * The name of the field whose value is being updated (e.g., product, category, model)
   */
  fieldName: string;
  /**
   * The value to set in the state.
   */
  fieldValue: string;
}

export const registerCaseField = createAction(
  'caseAssist/caseField/register',
  (payload: SetCaseFieldActionCreatorPayload) =>
    validatePayload(payload, {
      fieldName: requiredNonEmptyString,
      fieldValue: requiredEmptyAllowedString,
    })
);

export const updateCaseField = createAction(
  'caseAssist/caseField/update',
  (payload: SetCaseFieldActionCreatorPayload) =>
    validatePayload(payload, {
      fieldName: requiredNonEmptyString,
      fieldValue: requiredEmptyAllowedString,
    })
);

export interface FetchClassificationsThunkReturn {
  /** The successful classifications response. */
  response: GetCaseClassificationsResponse;
}

export type StateNeededByFetchClassifications = ConfigurationSection &
  CaseAssistConfigurationSection &
  CaseFieldSection &
  CaseInputSection &
  DebugSection;

export const fetchCaseClassifications = createAsyncThunk<
  FetchClassificationsThunkReturn,
  void,
  AsyncThunkCaseAssistOptions<StateNeededByFetchClassifications>
>(
  'caseAssist/classifications/fetch',
  async (_, {getState, rejectWithValue, extra: {apiClient}}) => {
    const state = getState();

    const fetched = await apiClient.getCaseClassifications(
      await buildFetchClassificationRequest(state)
    );

    if (isErrorResponse(fetched)) {
      return rejectWithValue(fetched.error);
    }

    return {
      response: fetched.success,
    };
  }
);
export const buildFetchClassificationRequest = async (
  state: StateNeededByFetchClassifications
): Promise<GetCaseClassificationsRequest> => ({
  accessToken: state.configuration.accessToken,
  organizationId: state.configuration.organizationId,
  url: state.configuration.platformUrl,
  caseAssistId: state.caseAssistConfiguration.caseAssistId,
  ...(state.configuration.analytics.enabled && {
    clientId: await getVisitorID(state.configuration.analytics),
  }),
  fields: state.caseInput,
  context: state.caseField
    ? prepareContextFromFields(state.caseField.fields)
    : undefined,
  locale: state.caseAssistConfiguration.locale,
  debug: state.debug,
});
