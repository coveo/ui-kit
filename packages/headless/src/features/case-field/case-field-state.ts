import {CaseAssistAPIErrorStatusResponse} from '../../api/service/case-assist/case-assist-api-client.js';

export const getCaseFieldInitialState = (): CaseFieldState => ({
  status: {
    loading: false,
    error: null,
    lastResponseId: '',
  },
  fields: {},
});

export interface CaseFieldStatus {
  /**
   * `true` if a request is in progress and `false` otherwise.
   */
  loading: boolean;
  /**
   * The Case Assist API error response.
   */
  error: CaseAssistAPIErrorStatusResponse | null;
  /**
   * The ID of the response.
   */
  lastResponseId: string;
}

export interface CaseFieldSuggestion {
  /**
   * The ID of the suggestion.
   */
  id: string;
  /**
   * The suggested field value.
   */
  value: string;
  /**
   * The confidence value of the suggestion in decimal.
   */
  confidence: number;
}

export interface CaseField {
  /**
   * The field to predict.
   */
  value: string;
  /**
   * The field value suggestions returned from the Coveo Customer Service API.
   */
  suggestions: CaseFieldSuggestion[];
}

export interface CaseFieldState {
  /**
   * The status of the case fields classification request.
   */
  status: CaseFieldStatus;
  /**
   * The case fields for which classifications are retrieved.
   */
  fields: {
    [fieldName: string]: CaseField;
  };
}
