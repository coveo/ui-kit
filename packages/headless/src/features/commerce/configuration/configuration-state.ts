import {CoreConfigurationState} from '../../configuration/configuration-state.js';

export interface ConfigurationState extends CoreConfigurationState {
  /**
   * The global headless engine Commerce API configuration.
   */
  commerce: CommerceState;
}

interface CommerceState {
  /**
   * The Commerce API base URL to use.
   *
   * By default, will append `/rest/organizations/{organizationId}/commerce/v2` to the automatically resolved
   * platform [organization endpoint](https://docs.coveo.com/en/mcc80216)
   * (i.e., `https;://<ORG_ID>.org<hipaa|dev|stg|>.coveo.com`)
   *
   * If necessary, you can override this value by specifying a `commerce.proxyBaseUrl` in the configuration of your
   * commerce engine, or by manually dispatching the `updateProxyBaseUrl` action.
   */
  apiBaseUrl?: string;
}

export const getConfigurationInitialState: () => ConfigurationState = () => ({
  accessToken: '',
  environment: 'prod',
  organizationId: '',
  analytics: {
    enabled: true,
    trackingId: '',
    source: {},
  },
  commerce: {},
});
