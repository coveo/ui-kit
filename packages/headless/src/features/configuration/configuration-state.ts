import {platformUrl} from '../../api/platform-client';
import {IRuntimeEnvironment} from 'coveo.analytics';

export interface ConfigurationState {
  /**
   * The unique identifier of the target Coveo Cloud organization (e.g., `mycoveocloudorganizationg8tp8wu3`)
   */
  organizationId: string;
  /**
   * The access token to use to authenticate requests against the Coveo Cloud endpoints. Typically, this will be an API key or search token that grants the privileges to execute queries and push usage analytics data in the target Coveo Cloud organization.
   */
  accessToken: string;
  /**
   * The Plaform URL to use.
   * By default, https://platform.cloud.coveo.com
   */
  platformUrl: string;
  /**
   * The global headless engine Search API configuration.
   */
  search: {
    /**
     * The Search API base URL to use.
     * By default, will append /rest/search/v2 to the platformUrl value.
     */
    apiBaseUrl: string;
    /**
     * The locale of the current user. Must comply with IETF’s BCP 47 definition: https://www.rfc-editor.org/rfc/bcp/bcp47.txt.
     */
    locale: string;
  };
  /**
   * The global headless engine Usage Analytics API configuration.
   */
  analytics: {
    /**
     * Specifies if analytics tracking should be enabled. By default analytics events are tracked.
     */
    enabled: boolean;
    /**
     * The Analytics API base URL to use.
     * By default, will append /rest/ua to the platformUrl value.
     */
    apiBaseUrl: string;
    /**
     * Origin level 2 is a usage analytics event metadata whose value should typically be the name/identifier of the tab from which the usage analytics event originates.
     *
     * When logging a Search usage analytics event, originLevel2 should always be set to the same value as the corresponding tab (parameter) Search API query parameter so Coveo Machine Learning models function properly, and usage analytics reports and dashboards are coherent.
     *
     * This value is optional, and will automatically try to resolve itself from the tab search parameter.
     */
    originLevel2: string;

    /**
     * Origin level 3 is a usage analytics event metadata whose value should typically be the URL of the page that linked to the search interface that’s making the request.
     *
     * When logging a Search usage analytics event, originLevel3 should always be set to the same value as the corresponding referrer Search API query parameter so usage analytics reports and dashboards are coherent.
     *
     * This value is optional, and will automatically try to resolve itself from the referrer search parameter.
     */
    originLevel3: string;
    /**
     * Optional analytics runtime environment, this is needed for analytics to work correctly if you're running outside of a browser.
     * See https://github.com/coveo/coveo.analytics.js for more info
     */
    runtimeEnvironment?: IRuntimeEnvironment;
  };
}

export const searchAPIEndpoint = '/rest/search/v2';
export const analyticsAPIEndpoint = '/rest/ua';

export const getConfigurationInitialState: () => ConfigurationState = () => ({
  organizationId: '',
  accessToken: '',
  platformUrl: platformUrl(),
  search: {
    apiBaseUrl: `${platformUrl()}${searchAPIEndpoint}`,
    locale: 'en-US',
  },
  analytics: {
    enabled: true,
    apiBaseUrl: `${platformUrl()}${analyticsAPIEndpoint}`,
    originLevel2: 'default',
    originLevel3: 'default',
  },
});
