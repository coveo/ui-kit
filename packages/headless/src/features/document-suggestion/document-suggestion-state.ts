import {CaseAssistAPIErrorStatusResponse} from '../../api/service/case-assist/case-assist-api-client.js';
import {DocumentSuggestionResponse} from '../../api/service/case-assist/get-document-suggestions/get-document-suggestions-response.js';

export const getDocumentSuggestionInitialState =
  (): DocumentSuggestionState => ({
    status: {
      loading: false,
      error: null,
      lastResponseId: '',
    },
    documents: [],
  });

export interface DocumentSuggestionStatus {
  /**
   * `true` if a request is in progress and `false` otherwise.
   */
  loading: boolean;
  /**
   * The Case Assist API error response.
   */
  error: CaseAssistAPIErrorStatusResponse | null;
  /**
   * The ID of the response.
   */
  lastResponseId: string;
}

export interface DocumentSuggestionState {
  /**
   * The status of the document suggestions request.
   */
  status: DocumentSuggestionStatus;
  /**
   * The retrieved document suggestions.
   */
  documents: DocumentSuggestionResponse[];
}
