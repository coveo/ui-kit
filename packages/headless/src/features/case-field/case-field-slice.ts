import {createReducer} from '@reduxjs/toolkit';
import {
  fetchCaseClassifications,
  registerCaseField,
  updateCaseField,
} from './case-field-actions.js';
import {getCaseFieldInitialState} from './case-field-state.js';

export const caseFieldReducer = createReducer(
  getCaseFieldInitialState(),

  (builder) => {
    builder
      .addCase(registerCaseField, (state, action) => {
        const {fieldName, fieldValue} = action.payload;
        state.fields[fieldName] = {
          value: fieldValue,
          suggestions: [],
        };
      })
      .addCase(updateCaseField, (state, action) => {
        const {fieldName, fieldValue} = action.payload;
        state.fields[fieldName].value = fieldValue;
      })
      .addCase(fetchCaseClassifications.rejected, (state, action) => {
        state.status.error = action.payload ?? null;
        state.status.loading = false;
      })
      .addCase(fetchCaseClassifications.fulfilled, (state, action) => {
        const defaultField = {value: '', suggestions: []};
        Object.entries(action.payload.response.fields).forEach(
          ([fieldName, content]) => {
            if (!state.fields[fieldName]) {
              state.fields[fieldName] = {...defaultField};
            }

            state.fields[fieldName].suggestions = content.predictions;
          }
        );
        state.status.lastResponseId = action.payload.response.responseId;
        state.status.error = null;
        state.status.loading = false;
      })
      .addCase(fetchCaseClassifications.pending, (state) => {
        state.status.loading = true;
      });
  }
);
