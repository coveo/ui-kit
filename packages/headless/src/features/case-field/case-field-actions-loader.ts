import {AsyncThunkAction, PayloadAction} from '@reduxjs/toolkit';
import {AsyncThunkCaseAssistOptions} from '../../api/service/case-assist/case-assist-api-client';
import {CaseAssistEngine} from '../../app/case-assist-engine/case-assist-engine';
import {caseFieldReducer as caseField} from '../../features/case-field/case-field-slice';
import {
  fetchCaseClassifications,
  FetchClassificationsThunkReturn,
  updateCaseField,
  SetCaseFieldActionCreatorPayload,
  StateNeededByFetchClassifications,
  registerCaseField,
} from './case-field-actions';

export type {SetCaseFieldActionCreatorPayload};

/**
 * The case field action creators.
 */
export interface CaseFieldActionCreators {
  /**
   * Registers a case field with the specified field and value.
   *
   * @param payload - The action creator payload.
   * @returns A dispatchable action.
   */
  registerCaseField(
    payload: SetCaseFieldActionCreatorPayload
  ): PayloadAction<SetCaseFieldActionCreatorPayload>;

  /**
   * Updates the specified case field with the provided value.
   *
   * @param payload - The action creator payload.
   * @returns A dispatchable action.
   */
  updateCaseField(
    payload: SetCaseFieldActionCreatorPayload
  ): PayloadAction<SetCaseFieldActionCreatorPayload>;

  /**
   * Retrieves the case classifications from the platform.
   * A single call retrieves classifications for all fields specified in the Case Assist configuration.
   * Case classifications are retrieved based on the case information entered so far.
   *
   * @returns A dispatchable action.
   */
  fetchCaseClassifications(): AsyncThunkAction<
    FetchClassificationsThunkReturn,
    void,
    AsyncThunkCaseAssistOptions<StateNeededByFetchClassifications>
  >;
}

/**
 * Loads the `case fields` reducer and returns possible action creators.
 *
 * @param engine - The headless engine.
 * @returns An object holding the action creators.
 */
export function loadCaseFieldActions(
  engine: CaseAssistEngine
): CaseFieldActionCreators {
  engine.addReducers({caseField});

  return {
    registerCaseField,
    updateCaseField,
    fetchCaseClassifications,
  };
}
