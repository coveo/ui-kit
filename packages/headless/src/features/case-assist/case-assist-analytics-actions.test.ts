import {createRelay} from '@coveo/relay';
import {ThunkExtraArguments} from '../../app/thunk-extra-arguments';
import {buildMockCaseAssistState} from '../../test/mock-case-assist-state';
import {
  buildMockCaseAssistEngine,
  MockedCaseAssistEngine,
} from '../../test/mock-engine-v2';
import {getCaseFieldInitialState} from '../case-field/case-field-state';
import {getConfigurationInitialState} from '../configuration/configuration-state';
import {getDocumentSuggestionInitialState} from '../document-suggestion/document-suggestion-state';
import {
  logCaseStart,
  logCaseNextStage,
  logCreateCase,
  logSolveCase,
  logAbandonCase,
  logUpdateCaseField,
  logAutoSelectCaseField,
  logClassificationClick,
  logDocumentSuggestionClick,
  logQuickviewDocumentSuggestionClick,
  logDocumentSuggestionOpen,
  logDocumentSuggestionRating,
} from './case-assist-analytics-actions';

const mockLogEnterInterface = jest.fn();
const mockLogMoveToNextCaseStep = jest.fn();
const mockLogCaseCreated = jest.fn();
const mockLogCaseSolved = jest.fn();
const mockLogCaseCancelled = jest.fn();
const mockLogUpdateCaseField = jest.fn();
const mockLogSelectFieldSuggestion = jest.fn();
const mockLogSelectDocumentSuggestion = jest.fn();
const mockLogRateDocumentSuggestion = jest.fn();

const emit = jest.fn();

jest.mock('@coveo/relay');

jest.mocked(createRelay).mockReturnValue({
  emit,
  getMeta: jest.fn(),
  on: jest.fn(),
  off: jest.fn(),
  updateConfig: jest.fn(),
  clearStorage: jest.fn(),
  version: 'foo',
});

jest.mock('coveo.analytics', () => {
  const mockCaseAssistClient = jest.fn(() => ({
    disable: jest.fn(),
    logEnterInterface: mockLogEnterInterface,
    logMoveToNextCaseStep: mockLogMoveToNextCaseStep,
    logCaseCreated: mockLogCaseCreated,
    logCaseSolved: mockLogCaseSolved,
    logCaseCancelled: mockLogCaseCancelled,
    logUpdateCaseField: mockLogUpdateCaseField,
    logSelectFieldSuggestion: mockLogSelectFieldSuggestion,
    logSelectDocumentSuggestion: mockLogSelectDocumentSuggestion,
    logRateDocumentSuggestion: mockLogRateDocumentSuggestion,
  }));

  return {
    CaseAssistClient: mockCaseAssistClient,
    history: {HistoryStore: jest.fn()},
  };
});

const exampleSubject = 'example subject';
const exampledDescription = 'example description';
const exampleCategory = 'example category';
const exampleProductId = 'abc';
const exampleId = '456';
const exampleFieldValue = 'bar';
const exampleClassificationValue = 'foo';
const exampleClassificationId = '123';
const exampleConfidence = 1;
const exampleClassificationResponseId = 'baz';
const exampleField = 'exampleField';
const exampleDocumentSuggestionResponseId = 'qux';
const exampleUniqueId = 'exampleUniqueId';
const exampleTitle = 'exampleTitle';
const exampleClickUri = 'exampleClickUri';
const exampleExcerpt = 'exampleExcerpt';
const exampleUri = 'exampleUri';
const exampleUrihash = 'exampleUrihash';

const exampleCaseInputState = {
  subject: {value: exampleSubject},
  description: {value: exampledDescription},
  category: {value: exampleCategory},
  productId: {value: exampleProductId},
  id: {value: exampleId},
};

const exampleCaseFieldState = {
  status: {
    ...getCaseFieldInitialState().status,
    lastResponseId: exampleClassificationResponseId,
  },
  fields: {
    [exampleField]: {
      value: exampleFieldValue,
      suggestions: [
        {
          value: exampleClassificationValue,
          id: exampleClassificationId,
          confidence: exampleConfidence,
        },
      ],
    },
  },
};

const exampleDocumentSuggestionState = {
  status: {
    ...getDocumentSuggestionInitialState().status,
    lastResponseId: exampleDocumentSuggestionResponseId,
  },
  documents: [
    {
      uniqueId: exampleUniqueId,
      title: exampleTitle,
      clickUri: exampleClickUri,
      hasHtmlVersion: true,
      excerpt: exampleExcerpt,
      fields: {
        uri: exampleUri,
        urihash: exampleUrihash,
      },
    },
  ],
};

describe('generated answer insight analytics actions', () => {
  let engine: MockedCaseAssistEngine;

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('when analyticsMode is `legacy`', () => {
    const expectedTicketPayload = {
      subject: exampleSubject,
      description: exampledDescription,
      category: exampleCategory,
      productId: exampleProductId,
      id: exampleId,
      custom: {
        exampleField: 'bar',
      },
    };

    const expectedClassificationSuggestion = {
      classificationId: exampleClassificationId,
      responseId: exampleClassificationResponseId,
      fieldName: exampleField,
      classification: {
        value: exampleClassificationValue,
        confidence: exampleConfidence,
      },
    };

    const expectedDocumentSuggestion = {
      suggestionId: exampleUniqueId,
      responseId: exampleDocumentSuggestionResponseId,
      suggestion: {
        documentUri: exampleUri,
        documentUriHash: exampleUrihash,
        documentTitle: exampleTitle,
        documentUrl: exampleClickUri,
        documentPosition: 1,
      },
    };

    beforeEach(() => {
      engine = buildMockCaseAssistEngine(
        buildMockCaseAssistState({
          configuration: {
            ...getConfigurationInitialState(),
            analytics: {
              ...getConfigurationInitialState().analytics,
              analyticsMode: 'legacy',
            },
          },
          caseInput: exampleCaseInputState,
          caseField: exampleCaseFieldState,
          documentSuggestion: exampleDocumentSuggestionState,
        })
      );
    });

    it('should log #logCaseStart with the right payload', async () => {
      await logCaseStart()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockLogEnterInterface;

      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse).toHaveBeenCalledWith({ticket: expectedTicketPayload});
    });

    it('should log #logCaseNextStage with the right payload', async () => {
      const exampleStageName = 'foo';
      await logCaseNextStage({stageName: exampleStageName})()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockLogMoveToNextCaseStep;

      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse).toHaveBeenCalledWith({
        ticket: expectedTicketPayload,
        stage: exampleStageName,
      });
    });

    it('should log #logCreateCase with the right payload', async () => {
      await logCreateCase()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockLogCaseCreated;

      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse).toHaveBeenCalledWith({ticket: expectedTicketPayload});
    });

    it('should log #logSolveCase with the right payload', async () => {
      await logSolveCase()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockLogCaseSolved;

      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse).toHaveBeenCalledWith({ticket: expectedTicketPayload});
    });

    it('should log #logAbandonCase with the right payload', async () => {
      await logAbandonCase()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockLogCaseCancelled;

      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse).toHaveBeenCalledWith({ticket: expectedTicketPayload});
    });

    it('should log #logUpdateCaseField with the right payload', async () => {
      await logUpdateCaseField(exampleField)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockLogUpdateCaseField;

      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse).toHaveBeenCalledWith({
        ticket: expectedTicketPayload,
        fieldName: exampleField,
      });
    });

    it('should log #logAutoSelectCaseField with the right payload', async () => {
      await logAutoSelectCaseField(exampleClassificationId)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockLogSelectFieldSuggestion;

      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse).toHaveBeenCalledWith({
        ticket: expectedTicketPayload,
        suggestion: {
          ...expectedClassificationSuggestion,
          autoSelection: true,
        },
      });
    });

    it('should log #logClassificationClick with the right payload', async () => {
      await logClassificationClick(exampleClassificationId)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockLogSelectFieldSuggestion;

      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse).toHaveBeenCalledWith({
        ticket: expectedTicketPayload,
        suggestion: expectedClassificationSuggestion,
      });
    });

    it('should log #logDocumentSuggestionClick with the right payload', async () => {
      await logDocumentSuggestionClick(exampleUniqueId)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockLogSelectDocumentSuggestion;

      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse).toHaveBeenCalledWith({
        ticket: expectedTicketPayload,
        suggestion: expectedDocumentSuggestion,
      });
    });

    it('should log #logQuickviewDocumentSuggestionClick with the right payload', async () => {
      await logQuickviewDocumentSuggestionClick(exampleUniqueId)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockLogSelectDocumentSuggestion;

      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse).toHaveBeenCalledWith({
        ticket: expectedTicketPayload,
        suggestion: {...expectedDocumentSuggestion, fromQuickview: true},
      });
    });

    it('should log #logDocumentSuggestionOpen with the right payload', async () => {
      await logDocumentSuggestionOpen(exampleUniqueId)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockLogSelectDocumentSuggestion;

      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse).toHaveBeenCalledWith({
        ticket: expectedTicketPayload,
        suggestion: {...expectedDocumentSuggestion, openDocument: true},
      });
    });

    it('should log #logDocumentSuggestionRating with the right payload', async () => {
      const exampleRatingValue = 1;
      await logDocumentSuggestionRating(exampleUniqueId, exampleRatingValue)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockLogRateDocumentSuggestion;

      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse).toHaveBeenCalledWith({
        ticket: expectedTicketPayload,
        suggestion: expectedDocumentSuggestion,
        rating: exampleRatingValue,
      });
    });
  });

  describe('when analyticsMode is `next`', () => {
    beforeEach(() => {
      engine = buildMockCaseAssistEngine(
        buildMockCaseAssistState({
          configuration: {
            ...getConfigurationInitialState(),
            analytics: {
              ...getConfigurationInitialState().analytics,
              analyticsMode: 'next',
            },
          },
          caseInput: exampleCaseInputState,
          caseField: exampleCaseFieldState,
          documentSuggestion: exampleDocumentSuggestionState,
        })
      );
    });

    it('should log #logCaseStart with the right payload', async () => {
      await logCaseStart()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logCreateCase with the right payload', async () => {
      await logCreateCase()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logSolveCase with the right payload', async () => {
      await logSolveCase()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logAbandonCase with the right payload', async () => {
      await logAbandonCase()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logUpdateCaseField with the right payload', async () => {
      await logUpdateCaseField(exampleField)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logAutoSelectCaseField with the right payload', async () => {
      await logAutoSelectCaseField(exampleClassificationId)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logClassificationClick with the right payload', async () => {
      await logClassificationClick(exampleClassificationId)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logDocumentSuggestionClick with the right payload', async () => {
      await logDocumentSuggestionClick(exampleUniqueId)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logQuickviewDocumentSuggestionClick with the right payload', async () => {
      await logQuickviewDocumentSuggestionClick(exampleUniqueId)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logDocumentSuggestionOpen with the right payload', async () => {
      await logDocumentSuggestionOpen(exampleUniqueId)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logDocumentSuggestionRating with the right payload', async () => {
      await logDocumentSuggestionRating(exampleUniqueId, 1)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });
  });
});
