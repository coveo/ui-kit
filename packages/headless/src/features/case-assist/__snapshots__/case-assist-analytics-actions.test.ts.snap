// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logAbandonCase with the right payload 1`] = `
[
  "caseAssist.cancel",
  {
    "reason": "quit",
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logAutoSelectCaseField with the right payload 1`] = `
[
  "caseAssist.selectFieldClassification",
  {
    "autoselected": true,
    "fieldClassification": {
      "id": "123",
      "responseId": "baz",
    },
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logCaseStart with the right payload 1`] = `
[
  "caseAssist.start",
  {},
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logClassificationClick with the right payload 1`] = `
[
  "caseAssist.selectFieldClassification",
  {
    "autoselected": false,
    "fieldClassification": {
      "id": "123",
      "responseId": "baz",
    },
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logCreateCase with the right payload 1`] = `
[
  "caseAssist.createTicket",
  {
    "ticket": {
      "category": "example category",
      "description": "example description",
      "id": "456",
      "subject": "example subject",
      "system": "abc",
    },
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logDocumentSuggestionClick with the right payload 1`] = `
[
  "caseAssist.documentSuggestionClick",
  {
    "documentSuggestion": {
      "id": "exampleUniqueId",
      "responseId": "qux",
    },
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logDocumentSuggestionOpen with the right payload 1`] = `
[
  "itemClick",
  {
    "actionCause": "open",
    "itemMetadata": {
      "title": "exampleTitle",
      "uniqueFieldName": "uniqueId",
      "uniqueFieldValue": "exampleUniqueId",
      "url": "exampleClickUri",
    },
    "position": 1,
    "searchUid": "qux",
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logDocumentSuggestionRating with the right payload 1`] = `
[
  "caseAssist.documentSuggestionFeedback",
  {
    "documentSuggestion": {
      "id": "exampleUniqueId",
      "responseId": "qux",
    },
    "liked": true,
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logQuickviewDocumentSuggestionClick with the right payload 1`] = `
[
  "itemClick",
  {
    "actionCause": "preview",
    "itemMetadata": {
      "title": "exampleTitle",
      "uniqueFieldName": "uniqueId",
      "uniqueFieldValue": "exampleUniqueId",
      "url": "exampleClickUri",
    },
    "position": 1,
    "searchUid": "qux",
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logSolveCase with the right payload 1`] = `
[
  "caseAssist.cancel",
  {
    "reason": "solved",
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logUpdateCaseField with the right payload 1`] = `
[
  "caseAssist.updateField",
  {
    "fieldName": "exampleField",
    "fieldValue": "bar",
  },
]
`;
